src/Razor/test/RazorLanguage.Test/Legacy/RazorDirectivesTest.cs

```diff
@@ -906,5 +906,52 @@ public void Directives_ReservedWordInsideCodeBlock()
         {
             ParseDocumentTest("@{ class }");
         }
+
+        [Fact]
+        public void DirectiveDescriptor_UnderstandsAttributeTokens()
+        {
+            // Arrange
+            var descriptor = DirectiveDescriptor.CreateDirective(
+                "custom",
+                DirectiveKind.SingleLine,
+                b => b.AddAttributeToken());
+
+            // Act & Assert
+            ParseDocumentTest(@"
+@custom [Serializable]
+@custom [DllImport(""user32.dll"", SetLastError=false, ExactSpelling=false)]
+",
+                new[] { descriptor });
+        }
+
+        [Fact]
+        public void DirectiveDescriptor_AttributeToken_BalancesBrackets()
+        {
+            // Arrange
+            var descriptor = DirectiveDescriptor.CreateDirective(
+                "custom",
+                DirectiveKind.SingleLine,
+                b => b.AddAttributeToken());
+
+            // Act & Assert
+            ParseDocumentTest(@"
+@custom [SomeCustom(new int[] { 1, 2, 3 }
+",
+                new[] { descriptor });
+        }
+
+        [Fact]
+        public void DirectiveDescriptor_AttributeToken_ErrorsIfDoesNotStartWithOpenBracket()
+        {
+            // Arrange
+            var descriptor = DirectiveDescriptor.CreateDirective(
+                "custom",
+                DirectiveKind.SingleLine,
+                b => b.AddAttributeToken());
+
+            // Act & Assert
+            ParseDocumentTest("@custom Serializable]",
+                new[] { descriptor });
+        }
```

Comment:
What tests do we need for error cases?

Reply:
Not sure what you mean. It creates a `diag.txt` with the error.